pkg/util/strategicpatch/patch.go

```diff
@@ -0,0 +1,454 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package strategicpatch
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"reflect"
+	"sort"
+)
+
+// An alternate implementation of JSON Merge Patch
+// (https://tools.ietf.org/html/rfc7386) which supports the ability to annotate
+// certain fields with metadata that indicates whether the elements of JSON
+// lists should be merged or replaced.
+//
+// For more information, see the PATCH section of docs/api-conventions.md.
+func StrategicMergePatchData(original, patch []byte, dataStruct interface{}) ([]byte, error) {
+	var o map[string]interface{}
+	err := json.Unmarshal(original, &o)
+	if err != nil {
+		return nil, err
+	}
+
+	var p map[string]interface{}
+	err = json.Unmarshal(patch, &p)
+	if err != nil {
+		return nil, err
+	}
+
+	t := reflect.TypeOf(dataStruct)
+	if t.Kind() == reflect.Ptr {
+		t = t.Elem()
+	}
+	if t.Kind() != reflect.Struct {
+		return nil, fmt.Errorf("strategic merge patch needs a struct, %s received instead", t.Kind().String())
+	}
+
+	result, err := mergeMap(o, p, t)
+	if err != nil {
+		return nil, err
+	}
+
+	return json.Marshal(result)
+}
+
+const specialKey = "__patch"
+
+// Merge fields from a patch map into the original map.
+func mergeMap(original, patch map[string]interface{}, t reflect.Type) (map[string]interface{}, error) {
+	// If the map contains "__patch: strict", don't merge it, just use the
+	// patch map directly. Later on, can add a non-recursive strict that only
+	// affects the map that the __patch is in.
+	if v, ok := patch[specialKey]; ok {
+		if v == "strict" {
+			delete(patch, specialKey)
+			return patch, nil
+		}
+	}
+
+	// Start merging the patch into the original.
+	for k, patchV := range patch {
+		// If the value of this key is null, delete the key if it exists in the
+		// original. Otherwise, skip it.
+		if patchV == nil {
+			if _, ok := original[k]; ok {
+				delete(original, k)
+			}
+			continue
+		}
+
+		_, ok := original[k]
+		if !ok {
+			// If it's not in the original document, just take the patch value.
+			if original == nil {
+				original = map[string]interface{}{}
+			}
+			original[k] = patchV
+		} else {
+			// If they're both maps or lists, recurse into the value.
+			// First find the fieldPatchStrategy and fieldPatchMergeKey.
+			fieldType, fieldPatchStrategy, fieldPatchMergeKey, err := lookupPatchMetadata(t, k)
+			if err != nil {
+				return nil, err
+			}
+
+			originalType := reflect.TypeOf(original[k])
+			patchType := reflect.TypeOf(patchV)
+			if originalType == patchType {
+				if originalType.Kind() == reflect.Map && fieldPatchStrategy != "replace" {
```

Comment:
Also, it's possible to override merge with strict, but not possible to override replace with merge?